### ISSUE TITLE
Collect build time statistics

### ISSUE THREAD

**Issue by lheckemann** [NixOS/nix#1710]
I've often found myself wanting to know how long a derivation took to build. The closest I've found to this has been using `stat /nix/var/log/nix/drvs/xx/xx[…]-foo-bar.drv` and subtracting the atime from the mtime. This is very brittle, however, and I think it would be nice for nix to record this information somewhere. Does this perhaps already exist? Any thoughts on the idea or implementation?

**Comment by danbst** [comment]
Hydra does collect build duration, but how to use it? 
https://hydra.nixos.org/build/90102549

> Duration: | 17s

But we don't know number of cores, parallelism and CPU frequency of builder machine. Maybe `user-time` collection would be great here, as building is usually CPU intensive, and other subsystems are not very loaded.

Local collector would be also great, for both successful builds and failed ones.

**Comment by vcunat** [comment]
Yes, collecting stats like user+sys time is the best suggestion so far.  I doubt there's something noticeably better that's easy to do, at least in the way Hydra.nixos.org looks (quite diverse).

**Comment by Gaelan** [comment]
Some other statistics that might be useful:

* max disk usage (might be able to get away with only measuring disk usage at the end, but maybe some build systems delete intermediate files before the end? in that case, we'd need to regularly measure)
  * would allow us to warn and/or offer to run GC if there's insufficient space to build something
* max memory usage
* lines/bytes of build output
  * assuming that builds spew out logs at a roughly constant rate, this would let us make a decent guess at an ETA even on machines faster/slower than Hydra

**Comment by timokau** [comment]
I love the last idea! To be a bit more precise, you could actually annotate every output line with a timestamp and then with the benefit of hindsight say "after 10% of the build time the builder hat produced 1000 lines". That would cover the case where for example the actual build is relatively quiet but the install phase spews a lot of information.

**Comment by zimbatm** [comment]
It would be very useful if that information was attached to the .narinfo file.

**Comment by anka-213** [comment]
I'm still interested in this happening

**Comment by ncfavier** [comment]
Annoying bot; this is important.

**Comment by lheckemann** [comment]
This could probably be bolted on using the `pre-build-hook` and `post-build-hook` options.

**Comment by lheckemann** [comment]
@fricklerhandwerk I don't see how this is related to error messages?

**Comment by fricklerhandwerk** [comment]
Probably a random mistake, I went through issues fairly quickly.

**Comment by yodaldevoid** [comment]
> This could probably be bolted on using the `pre-build-hook` and `post-build-hook` options.

I have just tried to do this, but it looks like `pre-build-hook` is currently not run for non-sandboxed builds (see https://github.com/NixOS/nix/issues/2985), just like the builds in docker containers that I am trying to measure.
